File: projects/punk0/code/draw3.js
// the next player draws list[0] cards
// and returns a copy of the list:
//   [1,2,3,4]
// returns
//   [1,2,3,4]
// prints:
//   draw 1 cards
function draw(x) {
  if (x.length > 0) {
    let n = x[0];
    if (n == 0) {
      console.log("next player skips");
    } else {
      console.log(`draw ${n} cards`);
    }
  }
  // the slice(start,end) method makes a
  // shallow copy of the array from
  // start up to end, if start and end
  // are ommited, it will return a
  // shallow copy of the whole array
  let r = x.slice();

  return r;
}

console.log(JSON.stringify(draw([1, 1, 2, 3, 3, 4, 1, 2, 7, 1])));

File: projects/punk0/code/draw3.go
package main

import (
	"fmt"
)

// the next player draws list[0] cards
// and returns a copy of the list:
//   [1,2,3,4]
// returns
//   [1,2,3,4]
// prints:
//   draw 1 cards
func draw(x []int) []int {
	if len(x) > 0 {
		n := x[0]
		if n == 0 {
			fmt.Printf("next player skips\n")
		} else {
			fmt.Printf("draw %d cards\n", n)
		}

	}
	// make a slice with length 0
	// and capacity len(x), so
	// len(x) ints are preallocated
	r := make([]int, 0, len(x))
	for _, v := range x {
		// while there is capacity
		// it will just add to the end
		// without reallocating
		r = append(r, v)
	}

	return r
}

func main() {
	fmt.Printf("%v\n", draw([]int{1, 1, 2, 3, 3, 4, 1, 2, 7, 1}))
}
